CARBON-16084 : Carbon C5 - Global Configuration Model

daneshk opened the pull request:
This enhancement is to,
1. Enable global configuration file(e.g: deployment.yaml) cross the product. such that,
- All the user configurations must have default values and it define in respective bean classes.
- Any default configuration can overridden by the configuration added in the global configuration file(e.g: deployment.yaml).
- Defining all configurations in the configuration file in not required. add only if we need to override the default value.
- If configurations are not specified in the configuration file, values defined in the bean class will apply by default.

2. Create configuration file (for documentation purposes) automatically by reading the configuration bean classes.

Changes made,
- Three new annotations (Configuration, Element, Ignore), added to core module.
- Maven plugin (ConfigDocumentMojo) to create configuration file, added as new module (plugins)
- Annotation Processor (ConfigurationProcessor) to discover configuration bean classes in the component, added as new module (processors)
- new OSGI service (ConfigProvider) to generate configuration object, added to core module. 
- new testcase (ConfigProviderImplTest.java) to test the osgi service.
- change existing testcases.
- remove carbon.yaml, deployment.properties.

daneshk committed:
carbon configuration module - phase 01

daneshk committed:
carbon configuration model: phase 02

daneshk committed:
change the format of Element annotation

daneshk committed:
add constant values and change the doc comment

daneshk committed:
configuration model: phase 03

daneshk committed:
configuration model: refactor the code

daneshk committed:
C5 configuration model: phase 04

daneshk committed:
C5 configuration model: phase 05

daneshk committed:
C5 configuration model: phase 06: remove carbon.yaml and deployment.properties

daneshk committed:
C5 configuration model: phase 07: refactor the code

daneshk committed:
formatting issue

daneshk committed:
refactor the code

this commented on core/src/main/java/org/wso2/carbon/kernel/configprovider/YAMLBasedConfigFileReader.java:
```diff
@@ -0,0 +1,97 @@
+    public Map<String, String> getDeploymentConfiguration() {
+        org.wso2.carbon.kernel.utils.Utils.checkSecurity();
+        File configFile = ConfigurationUtils.getConfigurationFileLocation(filename).toFile();
+        try (FileInputStream fileInputStream = new FileInputStream(configFile)) {
```
Instead of using old File, InputStream APIs, use NIO Path, Files APIs.

``` java
byte[] contentBytes = Files.readAllBytes(ConfigurationUtils.getConfigurationFileLocation(filename));
String content = new String(contentBytes, StandardCharsets.UTF_8));
```

daneshk replied:
will change it. thanks.

this commented on core/src/main/java/org/wso2/carbon/kernel/configprovider/YAMLBasedConfigFileReader.java:
```diff
@@ -0,0 +1,97 @@
+        } catch (IOException e) {
+            String errorMessage = "Failed populate Deployment Configuration from " + configFile.getName();
+            logger.error(errorMessage, e);
+            throw new RuntimeException(errorMessage, e);
```
Why throw RuntimeException?

daneshk replied:
We can change this to checked exception. thanks for pointing out

this commented on core/src/main/java/org/wso2/carbon/kernel/configprovider/YAMLBasedConfigFileReader.java:
```diff
@@ -0,0 +1,97 @@
+    private Map<String, String> getDeploymentConfigMap(String jsonString) {
+        Map<String, String> deploymentConfigs = new HashMap<>();
+        JSONObject jsonObject = new JSONObject(jsonString);
+        for (Object key : jsonObject.keySet()) {
```
Instead iterate through [Entry set](https://docs.oracle.com/javase/7/docs/api/java/util/Map.html#entrySet%28%29).

afkham replied:
Yes, can use foreach and then use Lambda

this commented on core/src/main/java/org/wso2/carbon/kernel/configprovider/utils/ConfigurationUtils.java:
```diff
@@ -0,0 +1,55 @@
+     * @return Path configuration file location
+     */
+    public static Path getConfigurationFileLocation(String filename) {
+        return Paths.get(org.wso2.carbon.kernel.utils.Utils.getCarbonConfigHome().toString(),
```
`Utils.getCarbonConfigHome()` retuns a Path, so you can use `Utils.getCarbonConfigHome().resolve(filename)` instead.

daneshk replied:
will change it. thanks.

imesh commented on core/src/main/java/org/wso2/carbon/kernel/annotations/Configuration.java:
```diff
@@ -0,0 +1,37 @@
+    String namespace() default NULL;
+    // field description, required
+    String description();
+    public static final String NULL = "NULL";
```
It might be better to explain the purpose of this constant using a comment.

imesh commented on core/src/main/java/org/wso2/carbon/kernel/config/model/CarbonConfiguration.java:
```diff
@@ -16,29 +16,39 @@
 
-    private String version = "5.0.0";
+    @Element(description = "server version")
+    private String version = "5.2.0-SNAPSHOT";
```
Would it be possible to set the version value at the build time by reading the version in pom.xml file? Otherwise, when this component being released, this value might need to be manually updated.

Niranjan-K replied:
@imesh : the actual value is retrieved from the yaml configuration file. IMO this version can remain "5" as the default value. WDYT?

Niranjan-K replied:
As per the offline discussion with @afkham, it was suggested that there should not be any value here and should be populated to the file during build time using the version from pom.
Also the yaml file created for this component should be packed with the corresponding feature, but it should not be mentioned in the feature's p2.inf. This is done so that the file is not copied to the product.

daneshk replied:
Created issue  #1192 for version value and fixed with commit 6a2d697ff94d65c53ce5f9b94f3b8c751b3cf531

imesh commented on core/src/main/java/org/wso2/carbon/kernel/configprovider/XMLBasedConfigFileReader.java:
```diff
@@ -0,0 +1,95 @@
+            String jsonString = ConfigResolverUtils.convertXMLToJSON(xmlFileString);
+            return getDeploymentConfigMap(jsonString);
+        } catch (IOException e) {
+            String errorMessage = "Failed populate Deployment Configuration from " + configFile.getName();
```
Do we need to make the first two letters of "Deployment Configuration" uppercase?

daneshk replied:
corrected it. thanks

imesh commented on core/src/main/java/org/wso2/carbon/kernel/configprovider/XMLBasedConfigFileReader.java:
```diff
@@ -0,0 +1,95 @@
+        } catch (IOException e) {
+            String errorMessage = "Failed populate Deployment Configuration from " + configFile.getName();
+            logger.error(errorMessage, e);
+            throw new RuntimeException(errorMessage, e);
```
As it looks like an IOException can occur in this scenario due to many reasons. However, the caller may not know it until the program is run. Therefore it might be better to let the method caller know that by throwing the IOException itself or a custom exception.

daneshk replied:
will change it to custom exception. thanks for pointing out

imesh commented on core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderImpl.java:
```diff
@@ -0,0 +1,241 @@
+            String jsonConfigString = deploymentConfigs.get(namespace);
+            String yamlConfigString = ConfigurationUtils.convertJSONToYAML(jsonConfigString);
+            if (logger.isDebugEnabled()) {
+                logger.debug("class name: " + configClass.getSimpleName() + " | new configurations : \n" +
```
Might need to remove the space before the colon: "configurations : \n"

daneshk replied:
changed it. thanks

imesh commented on core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderImpl.java:
```diff
@@ -0,0 +1,241 @@
+            try {
+                return configClass.newInstance();
+            } catch (InstantiationException | IllegalAccessException e) {
+                throw new CarbonConfigurationException("Error while creating configuration Instance : "
```
Might need to remove the space before the colon: "Instance :". Is there a reason to make the letter "I" uppercase?

daneshk replied:
changed it. thanks

imesh commented on core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderImpl.java:
```diff
@@ -0,0 +1,241 @@
+            Yaml yaml = new Yaml();
+            return yaml.loadAs(jsonProcessedString, Map.class);
+        }
+        logger.error("configuration doesn't exist for the namespace: " + namespace + " in deployment yaml. hence " +
```
The first letter after the period might need to be uppercase: "hence".

daneshk replied:
changed it. thanks

imesh commented on core/src/test/resources/yaml/conf/deployment.yaml:
```diff
@@ -0,0 +1,55 @@
+################################################################################
+#   Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
The year might need to be updated to 2016.

daneshk replied:
changed it. thanks for pointing out.

imesh commented on plugins/configuration-plugin/src/main/java/org/wso2/carbon/plugins/configuration/ConfigDocumentMojo.java:
```diff
@@ -0,0 +1,341 @@
+@Mojo(name = "create-doc")
+public class ConfigDocumentMojo extends AbstractMojo {
+
+    private static final Logger logger = LoggerFactory.getLogger(ConfigDocumentMojo.class.getName());
```
In most of the classes logger variable is named as log.

daneshk replied:
@imesh in carbon-kernel, logger variable is named as logger in all classes. So I followed the same convention.
Yes, in other repos it named as log.

daneshk committed:
changes made after reviewing the code

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/config/model/DeploymentConfig.java:
```diff
@@ -15,17 +15,32 @@
  * @since 5.0.0
  */
+@Configuration(description = "Deployment Engine related configurations")
 public class DeploymentConfig {
```
This class is not used in kernel anymore. It is now used with carbon-deployment. So can we check and remove these references from kernel code base?

Kishanthan commented on config-creator/pom.xml:
```diff
@@ -0,0 +1,63 @@
+            <plugin>
+                <groupId>org.wso2.carbon</groupId>
+                <artifactId>org.wso2.carbon.plugins.configuration</artifactId>
+                <version>${carbon.kernel.version}</version>
```
we don't need to define this as this version should come from parent

daneshk replied:
Corrected. Thanks

Kishanthan commented on config-creator/pom.xml:
```diff
@@ -0,0 +1,63 @@
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>config-creator</artifactId>
+    <name>WSO2 Carbon Kernel - CONFIGURATION CREATOR</name>
```
Can we not use "CONFIGURATION CREATOR" and instead use simple letters? Please look at the pattern used in other components.

daneshk replied:
Corrected. thanks

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/CarbonCoreComponent.java:
```diff
@@ -34,12 +32,11 @@ public void activate() {
 
             // 2) Creates the CarbonRuntime instance using the Carbon configuration provider.
-            CarbonRuntime carbonRuntime = CarbonRuntimeFactory.createCarbonRuntime(configProvider);
+             CarbonRuntime carbonRuntime = CarbonRuntimeFactory.createCarbonRuntime(configProvider);
```
This looks not formatted as there is an additional space added in front.

daneshk replied:
corrected. thanks

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/DataHolder.java:
```diff
@@ -81,4 +84,22 @@ public void setRuntimeManager(RuntimeManager runtimeManager) {
+     *
+     * @return optConfigProvider
+     */
+    public Optional<ConfigProvider> getOptConfigProvider() {
```
The naming convention is not correct. The return can be an optional but that does not need to be part of the method name.

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/DataHolder.java:
```diff
@@ -81,4 +84,22 @@ public void setRuntimeManager(RuntimeManager runtimeManager) {
+     *
+     * @param optConfigProvider
+     */
+    public void setOptConfigProvider(Optional<ConfigProvider> optConfigProvider) {
```
Why do we have to get and set a optional value directly. IMO, we can use the Optional return type with the getter method but for setter, you can directly use the value. Or do we have a proper scenario here to use optional?

daneshk replied:
No, we don't need optional. will remove it. thanks for pointing out

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderComponent.java:
```diff
@@ -0,0 +1,78 @@
+            unbind = "unRegisterSecureVault"
+    )
+    protected void registerSecureVault(SecureVault secureVault) {
+        ConfigProviderDataHolder.getInstance().setOptSecureVault(Optional.ofNullable(secureVault));
```
Once you fix the ConfigProviderDataHolder  API not at accept Optiona<> as values, this call also will change.

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderDataHolder.java:
```diff
@@ -0,0 +1,45 @@
+        return optSecureVault;
+    }
+
+    public void setOptSecureVault(Optional<SecureVault> optSecureVault) {
```
The pattern followed here is not correct as-well. Normally we set the values directly but, when returning we return an optional.

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderImpl.java:
```diff
@@ -0,0 +1,241 @@
+    private void loadDeploymentConfiguration(ConfigFileReader configFileReader) throws CarbonConfigurationException {
+        if (deploymentConfigs == null) {
+            synchronized (this) {
+                if (deploymentConfigs == null) {
```
Why do we need this synchronized check?

daneshk replied:
this method loads the configuration from the deployment.yaml on request if config map is null. if two or more requests comes at the same time when deploymentConfigs is null (first call), all request will read the file to load deployment configuration. to avoid that I added synchronized block.

daneshk committed:
remove optional<> from configProvider and secureVault, change the componant name according to the convension. change the formatting issues

daneshk committed:
remove deploymentConfig from carbonConfiguration

daneshk committed:
move annotation processor to core/annotations module, in order to reduce the extra dependency. add compiler arg to stop process execution in kernel

Niranjan-K commented on core/src/main/java/org/wso2/carbon/kernel/annotations/Configuration.java:
```diff
@@ -0,0 +1,39 @@
+import java.lang.annotation.Target;
+
+/**
+ * Configuration annotation for configuration tool. Class level annotation
```
It would be better to have the 2nd sentence in the next line

daneshk replied:
Corrected. thanks

Niranjan-K commented on core/src/main/java/org/wso2/carbon/kernel/annotations/Element.java:
```diff
@@ -0,0 +1,35 @@
+import java.lang.annotation.Target;
+
+/**
+ * Element annotation for configuration tool. Field level annotation
```
It would be better to have the 2nd sentence in the next line

daneshk replied:
Corrected. thanks

Niranjan-K commented on core/src/main/java/org/wso2/carbon/kernel/annotations/Ignore.java:
```diff
@@ -0,0 +1,31 @@
+import java.lang.annotation.Target;
+
+/**
+ * Ignore annotation for configuration tool, Field level annotation
```
The first line should end with a full stop else it will fail in Windows

daneshk replied:
Corrected.thanks

Niranjan-K commented on core/src/main/java/org/wso2/carbon/kernel/configprovider/YAMLBasedConfigFileReader.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    /**
+     * this method reads deployment.yaml file and return configuration map which is used for overriding default
```
It would be better to start with capital, i.e., "This"

daneshk replied:
Corrected. thanks

Niranjan-K commented on core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderImpl.java:
```diff
@@ -0,0 +1,242 @@
+
+
+    /**
+     * This method loads deployment configs in deployment.yaml. loads only if config
```
It seems that the explaining is missing :)

daneshk replied:
Corrected.thanks

Niranjan-K commented:
@daneshk : Have we tested building this in windows?

daneshk committed:
add new unit tests to increase the code coverity

daneshk committed:
corrected the comments to fixed the checkstyle error in windows, first line should be ended

imesh commented:
@daneshk Shall we finalize this and merge? I think there are few more comments above we might need to analyze. Thanks!

daneshk committed:
fixed java-doc comments issues

daneshk committed:
Merge branch 'config-model' of https://github.com/daneshk/carbon-kernel into config-model

Niranjan-K commented on core/src/test/resources/yaml/conf/deployment.yaml:
```diff
@@ -0,0 +1,44 @@
+wso2.carbon:
+  id: carbon-kernel
+  name: WSO2 Carbon Kernel
+  version: ${carbon.version}
```
These 3 values should be environment variables like ${server.key}, etc., so that it can be replaced during the build time using filter.properties.

daneshk replied:
This is for the unit test. I change those values in deployment.yaml which is packed in feature. commit revision 05850e1c5482fffbe4efb4c802650ea7ec54cac9

Niranjan-K commented on tests/osgi-tests/src/test/resources/carbon-context/deployment.yaml:
```diff
@@ -0,0 +1,53 @@
+    # server name
+  name: WSO2 Carbon Kernel
+    # server version
+  version: 5.2.0-SNAPSHOT
```
These values need to be removed and only the minimum configuration values should be there in deployment.yaml

daneshk replied:
version property is removed from the configuration file. commit revision 6a2d697ff94d65c53ce5f9b94f3b8c751b3cf531

daneshk committed:
Read Server Version from Project pom file and Ignore the version property from configuration file.

This commit is to remove the hardcoded default value in version property in CarbonConfiguration bean class and get the version value from the project pom file and also skip the version property from config docs.

Resolves: #1192

daneshk committed:
Pack config document for Carbon Configurations inside carbon core feature.

Generated config document from the CarbonConfiguration bean class was inside the jar file. This PR is to copy the config document to carbon core feature inside separate config-docs directory.

Resolves: #1193

daneshk committed:
Add Optional type validation to maven plugin when create config-docs

Optional field type validation needs to be added to maven plugin to generate correct configuration document. If the field default value is Optional.empty(), set the config value as null.

Resolves #1195

daneshk committed:
Change the default values of id and name in deployment.yaml

Change the default values of id and name in deployment.yaml to carbon-kernel and WSO2 Carbon Kernel instead of getting it from the server properties.

daneshk committed:
Remove server version property in sample deployment.yaml.

Server version property value was there in sample deployment.yaml which is used for testcases. This commit is to remove the version property from the all the configuration files in carbon-kernel repo.

daneshk committed:
Get field description from the configuration annotation of the nested classes

This commit is to read the description from the configuration annotation of the nested class instead of getting the description from the Element annotation of the reference field in the parent class. Since configuration annotation is required, there will description for the nested fields.
